### PR TITLE
Fix multiple closures on same line all resolving to the first one

### LINKED ISSUE #164: Multiple closures on the same line all resolve to the first one
## Description

When multiple closures are defined on the same source line with identical signatures, they all resolve to the first closure after a serialize/unserialize roundtrip. No error is thrown; the wrong closure silently executes.

## Reproduction

```php
<?php
require 'vendor/autoload.php';

use Opis\Closure\SerializableClosure;

$closures = [fn() => 'a', fn() => 'b', fn() => 'c'];

$serialized = array_map(function($c) {
    return serialize(new SerializableClosure($c));
}, $closures);

$results = array_map(function($s) {
    return unserialize($s)();
}, $serialized);

var_dump($results);
// Expected: ['a', 'b', 'c']
// Actual:   ['a', 'a', 'a']
```

## Root cause

Two places contribute:

1. **`AbstractParser::resolve()`** uses a cache key of `{prefix}/{fileKey}/{startLine}/{endLine}`. Multiple closures sharing the same line hit the same cache entry, so the second closure returns the first one's parsed info.

2. **`ClosureParser::findFunctionIndex()`** scans from `getStartLine()` and returns the first `T_FN`/`T_FUNCTION` token it finds. There is no mechanism to distinguish which closure on that line corresponds to which `\Closure` object.

## Impact

This affects any code that serializes arrays of inline closures, e.g. job chains, pipeline stages, callback arrays. The same bug exists in `laravel/serializable-closure` (issue laravel/serializable-closure#131) and a fix is in progress there.

Multi-line closures (each on their own line) are not affected.

## Environment

- opis/closure: 4.5.0
- PHP: 8.4.19
- OS: Windows 11

### PR DESCRIPTION
**Pull request by JoshSalway**
Fixes #164

## Context

This is a documented caveat (https://opis.io/closure/4.x/caveats.html) -- PHP reflection only provides line numbers, not column numbers, making it impossible to perfectly distinguish closures on the same line.

This PR proposes a possible workaround using a `WeakMap` to track candidate indices per closure object. It covers the common case (closures serialized in array order) but has a known limitation with out-of-order serialization (see below).

This is offered as a proof of concept. Whether the tradeoff is worth it is up to the maintainers to decide.

### Example 1: Data transform pipeline

**Before (bug):** Compact inline closures silently resolve to the wrong one:

```php
$transforms = [fn($x) => $x * 2, fn($x) => $x + 10, fn($x) => $x - 5];

$serialized = array_map(fn($t) => serialize(new SerializableClosure($t)), $transforms);
$restored = array_map(fn($s) => unserialize($s), $serialized);

$restored[0](10); // 20 ✓
$restored[1](10); // 20 ✗ expected 20 (x + 10)
$restored[2](10); // 20 ✗ expected 5  (x - 5)
// All three silently run the first transform (x * 2)
```

**After (fix):** Each closure preserves its identity:

```php
$restored[0](10); // 20 ✓ (x * 2)
$restored[1](10); // 20 ✓ (x + 10)
$restored[2](10); // 5  ✓ (x - 5)
```

---

### Example 2: Job chains

Job chains are particularly dangerous because wrong execution has real consequences:

```php
$steps = [fn() => ProcessPayment::run($order), fn() => SendConfirmation::run($order), fn() => UpdateInventory::run($order)];
```

**Before (bug):**

```
ProcessPayment::run($order)  ← runs ✓
ProcessPayment::run($order)  ← runs again (expected SendConfirmation) ✗
ProcessPayment::run($order)  ← runs again (expected UpdateInventory) ✗
// No error thrown. Payment charged 3x, confirmation never sent, inventory never updated.
```

**After (fix):**

```
ProcessPayment::run($order)    ✓
SendConfirmation::run($order)  ✓
UpdateInventory::run($order)   ✓
```

---

The workaround (separate lines) works but can't be enforced -- formatters, minifiers, and developers writing compact code will hit this silently.

## Real-world reports

- [frodeborli/serializor](https://github.com/frodeborli/serializor) (requires `opis/closure ^4.4`) independently discovered this bug (their [issue #12](https://github.com/frodeborli/serializor/issues/12)) and built explicit exception-throwing detection as a workaround, since opis/closure returns wrong results silently.
- [b7s/parallite-php](https://github.com/b7s/parallite-php) uses opis/closure for parallel closure execution. Their [`awaitAll(array $closures)`](https://github.com/b7s/parallite-php/blob/dc22e1ee3373ddd574df5fe1270248a986df8255/src/ParalliteClient.php#L203) accepts an array of closures, which are each serialized via [`\Opis\Closure\serialize()`](https://github.com/b7s/parallite-php/blob/dc22e1ee3373ddd574df5fe1270248a986df8255/src/Service/Parallite/SocketService.php#L201) in [`serializeClosure()`](https://github.com/b7s/parallite-php/blob/dc22e1ee3373ddd574df5fe1270248a986df8255/src/Service/Parallite/SocketService.php#L187-L209). Users passing inline closure arrays are directly exposed to this bug.

## Before/after test results

| | Master (before) | Fix branch (after) |
|---|---|---|
| Total tests | 183 | 189 (+6) |
| Passing | 179 | 185 |
| Errors | 4 (pre-existing) | 4 (same) |
| Same-line test | **FAIL** -- `a, a, a` | **PASS** -- `a, b, c` |

Zero regressions. The 4 pre-existing errors are identical on both branches.

## How it works

- `findFunctionIndex()` collects all closure tokens on the matching line range as candidates
- A static `WeakMap` keyed by the Closure object assigns each one a unique candidate index
- The candidate index is included in the cache key for proper disambiguation
- Single-closure and different-line paths are unchanged

## Known limitation

Candidate indices are assigned in serialization order, not source order. This means:

**Works (standard patterns):**
```php
$closures = [fn() => 'a', fn() => 'b', fn() => 'c'];
// Sequential processing -- all standard patterns work
array_map(fn($c) => serialize(new SerializableClosure($c)), $closures); // correct
foreach ($closures as $c) { serialize(new SerializableClosure($c)); }   // correct
```

**Breaks (deliberate out-of-order access):**
```php
$closures = [fn() => 'a', fn() => 'b', fn() => 'c'];
serialize(new SerializableClosure($closures[2])); // gets candidate 0 (wrong)
serialize(new SerializableClosure($closures[0])); // gets candidate 1 (wrong)
```

This last pattern was not found in any of 250+ public repos surveyed. The workaround remains the same as the current documented caveat: place closures on separate lines.

## Tests

6 new tests covering:
- Multiple arrow closures in array preserve identity
- Two closures on same line
- Closures on different lines (regression guard)
- Single closure unaffected
- Same-line closures with parameters
- Same-line static closures

## Impact on documented caveat

If merged, the [caveats page](https://opis.io/closure/4.x/caveats.html) could be updated from:

> **Current:** "Due to PHP's own limitations, the library is unable to distinguish between two or more closures when they are on the same line."

To:

> **Updated:** "Multiple closures on the same line now work when serialized in standard patterns (`array_map`, `foreach`, sequential iteration). Serializing same-line closures in a deliberately non-sequential order (e.g., accessing `$closures[2]` before `$closures[0]`) may still produce incorrect results. For guaranteed correctness in all cases, place each closure on its own line."

This narrows the caveat from "doesn't work" to "works in practice, edge case remains."

### CODE REVIEW

File: tests/PHP80/SameLineClosureTest.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        $this->assertEquals('b', $results[1]());
+        $this->assertEquals('c', $results[2]());
+    }
+
+    public function testTwoClosuresOnSameLine()
+    {
+        $a = fn() => 'x'; $b = fn() => 'y';
+
+        $ra = $this->process($a);
+        $rb = $this->process($b);
```
Comment (sorinsarca) [inline tests/PHP80/SameLineClosureTest.php:27]: Try to swap those two lines, the result should be the same if implementation is correct.

Comment (sorinsarca) [reply tests/PHP80/SameLineClosureTest.php:27]: When you change the order to 

```
$rb = $this->process($b);
$ra = $this->process($a);
```

the codes doesn't work, but it should work if correct.

**Comment by sorinsarca** [comment]
@JoshSalway read the review first.

This PR tries to solve #165 which currently cannot be solved.


If you have some ideas on how to solve it, please discuss there, I won't accept any other PR for this matter.

**Comment by JoshSalway** [comment]
@sorinsarca Thanks for your feedback, as mentioned earlier cant find people running into this issue in the wild so will have to stay as a known caveat.
